### ISSUE TITLE
Validation on DELETE action

### ISSUE THREAD

**Issue by CedricYii** [api-platform/api-platform#1928]
It looks like validation on DELETE action has been explicitly prevented in ValidateListener::onKernelView()
```php
//src\Validator\EventListener\ValidateListener.php::onKernelView()
        if (   
            $controllerResult instanceof Response   
            || $request->isMethodSafe()   
            || $request->isMethod('DELETE')  //<========================== here, why so much hate?   
            || !($attributes = RequestAttributesExtractor::extractAttributes($request))   
            || !$attributes['receive']   
            || $this->isOperationAttributeDisabled($attributes, self::OPERATION_ATTRIBUTE_KEY)   
        ) {   
            return;   
        }
```

However validation is very useful to enforce some business logic on the action itself (and share its behavior with the regular controller form view and the api platform).
Would it be possible to let validation occurs on DELETE action?
Alternatively add a parameter to enable it?

**Comment by maks-rafalko** [comment]
+1 for the feature request. In the meantime, we are doing validation via custom controller:

```
 *         "delete_business_owner"={
 *             "method"="DELETE",
 *             "controller"=DeleteProduct::class,
 *             "path"="/products/{id},
 *         },
```

and controller:

```php
final class DeleteProduct
{
    public function __invoke(Product $data): Response
    {
        $this->validator->validate($data);

        $this->em->remove($data);
        $this->em->flush();

        return new Response(null, Response::HTTP_NO_CONTENT);
    }
}
```

**Comment by CedricYii** [comment]
Thanks for sharing your workaround.
It's working great, however it means duplicating this code for every delete actions.
We tried to solve this hard-coded behavior by reversing it in an event subscriber:
```
<?php
//src\EventSubscriber\DeleteValidationSubscriber.php

namespace App\EventSubscriber;

use ApiPlatform\Core\Exception\ResourceClassNotFoundException;
use ApiPlatform\Core\Metadata\Resource\Factory\ResourceMetadataFactoryInterface;
use ApiPlatform\Core\Util\RequestAttributesExtractor;
use ApiPlatform\Core\Validator\ValidatorInterface;
use Symfony\Component\EventDispatcher\EventSubscriberInterface;
use Symfony\Component\HttpFoundation\Response;
use Symfony\Component\HttpKernel\Event\ViewEvent;
use Symfony\Component\HttpKernel\KernelEvents;

class DeleteValidationSubscriber implements EventSubscriberInterface
{
    private $validator;
    private $resourceMetadataFactory;

    /**
     * @param ValidatorInterface $validator
     * @param ResourceMetadataFactoryInterface $resourceMetadataFactory
     */
    public function __construct(ValidatorInterface $validator, ResourceMetadataFactoryInterface $resourceMetadataFactory)
    {
        $this->validator = $validator;
        $this->resourceMetadataFactory = $resourceMetadataFactory;
    }

    /**
     * Validates data returned by the controller for DELETE action because api-platform is ignoring it
     *
     * @param ViewEvent $event
     * @throws ResourceClassNotFoundException
     */
    public function onKernelView(ViewEvent $event)
    {
        $controllerResult = $event->getControllerResult();
        $request = $event->getRequest();

        if (
            $controllerResult instanceof Response
            || !$request->isMethod('DELETE')
            || $request->isMethodSafe()
            || !($attributes = RequestAttributesExtractor::extractAttributes($request))
            || !$attributes['receive']
        ) {
            return;
        }

        $resourceMetadata = $this->resourceMetadataFactory->create($attributes['resource_class']);

        $validationGroups = $resourceMetadata->getOperationAttribute($attributes, 'validation_groups', null, true);
        if (!is_null($validationGroups)) {
            $this->validator->validate($controllerResult, ['groups' => $validationGroups]);
        }
    }

    /**
     * @return array|string[]
     */
    public static function getSubscribedEvents()
    {
        return [
            KernelEvents::VIEW => ['onKernelView', 64],
        ];
    }
}
```
With that in place, validation is working as expected whenever "validation_groups" is explicitly defined on the "delete" item operation without other extra code.
Does it make sense?

**Comment by benblub** [comment]
The current recommendation would be to use DataPersister https://api-platform.com/docs/core/data-persisters/

```php
    use ApiPlatform\Core\Validator\ValidatorInterface;

    // ...

    /**
     * @param mixed $data
     */
    public function remove($data, array $context = []): void
    {
        /** @var User $data */
        $this->validator->validate($data, ['groups' => 'Delete']);

        $this->em->remove($data);
        $this->em->flush();
    }
```

**Comment by CedricYii** [comment]
Thanks for your indication.
Doing this way, the code would need to be duplicated for each action and the validation group spreaded away from the entity. That's really something we try to avoid.
And the validation error might not be sent back properly formated to the client.

**Comment by benblub** [comment]
Thats ok. In this approach the validation errors are send formatted back to the client. But fair enough there are different ways and you have to pick whats fit your Project. If you really need to validate every Entity for a delete request and rarely using DataPersisters at all -> go with Subscribers. 

If you have just 2-3 Entities where Validation is needed for an Entity and there are DataPersisters allready there then go with DataPersisters. (less Code, not called on every Request)

**Comment by CedricYii** [comment]
All right, thanks. I see what you mean. 
We are trying to avoid DataPersisters because we want to share most of the code with our frontend (and using ApiPlatform as a bundle). Good to know for the validation behavior though.
It would be nice to have the event subscriber to allow it directly, so there would have no overhead and things would be simpler, I might submit a PR.

**Comment by bogdankorobka** [comment]
Hello, have any news on this issue?

**Comment by soyuka** [comment]
yes this has been solved in 3.2/3.3, https://api-platform.com/docs/guides/delete-operation-with-validation/ 

on previous versions, create a provider and validate yourself the data
